### PR TITLE
Mnes 1098 add open id authentication to mnestix browser

### PR DESCRIPTION
**Pull request by pawel-baran-se**

### CODE REVIEW

File: .env.local [hunk 1]
```diff
@@ -9,5 +9,13 @@ DISCOVERY_API_URL: 'http://localhost:5064/discovery'
 REGISTRY_API_URL: 'http://localhost:8083'
 THEME_LOGO_MIME_TYPE: 'image/svg+xml'
 THEME_PRIMARY_COLOR: '#005962'
-THEME_SECONDARY_COLOR: '#147f8a'
-THEME_LOGO_URL: ''
\ No newline at end of file
+THEME_SECONDARY_COLOR: '#000000'
```
Comment (GailMelanie) [inline .env.local:12]:
```suggestion
THEME_SECONDARY_COLOR: '#147f8a'
```

File: src/lib/api/basyx-v3/api.ts [hunk 2]
```diff
@@ -133,7 +133,7 @@ export const AssetAdministrationShellRepositoryApiFp = function (configuration?:
                 configuration,
             ).getAssetAdministrationShellById(aasId, options);
             return async (fetch: FetchAPI = isomorphicFetch, basePath: string = BASE_PATH) => {
-                const response = await fetch(basePath + localVarFetchArgs.url, localVarFetchArgs.options);
+                const response = await fetch.fetch(basePath + localVarFetchArgs.url, localVarFetchArgs.options);
```
Comment (GailMelanie) [inline src/lib/api/basyx-v3/api.ts:136]: can one of the "fetch" be renamed?

File: src/lib/api/infrastructure.ts [hunk 3]
```diff
@@ -10,46 +10,37 @@ const initializeRequestOptions = async (bearerToken: string, init?: RequestInit)
...
+            const response = await fetch(url, await initializeRequestOptions(await getBearerToken(), init));
 
             if (response.status !== 401) {
                 return response;
             }
-            // Todo route to login page
-            // await redirectToLoginPage();
-
+            // If not logged in, redirect to sign-in
+            //await signIn('keycloak');
```
Comment (GailMelanie) [inline src/lib/api/infrastructure.ts:33]: can this be removed?

File: docker-compose/compose.dev.yml [hunk 4]
```diff
@@ -12,7 +12,6 @@ services:
       target: dev
     environment:
       REGISTRY_API_URL: 'https://registry.dpp40-2-v2.industrialdigitaltwin.org/'
-      SUBMODEL_REGISTRY_API_URL: 'http://localhost:8084'
```
Comment (GailMelanie) [inline docker-compose/compose.dev.yml:15]: why was the submodel registry removed?

File: src/authConfig.ts [hunk 5]
```diff
@@ -16,37 +19,81 @@ declare global {
...
-            authority, // This is a URL (e.g. https://login.microsoftonline.com/{your tenant ID})
-            redirectUri,
+const keycloakEnabled = process.env.KEYCLOAK_ENABLED?.toLowerCase() === 'true'.toLowerCase();
+
+export const authOptions: AuthOptions = {
+    providers: [
+        ...(keycloakEnabled ? [
+            KeycloakProvider({
+                clientId: process.env.KEYCLOAK_CLIENT_ID ? process.env.KEYCLOAK_CLIENT_ID : '',
+                clientSecret: 'process.env.KEYCLOAK_CLIENT_SECRET',
```
Comment (Xmilofranke) [inline src/authConfig.ts:29]: Why is this env variable hardcoded as a string?

File: src/authConfig.ts [hunk 6]
```diff
@@ -16,37 +19,81 @@ declare global {
...
+                issuer: process.env.KEYCLOAK_ISSUER
+            })
+        ] : [
+            AzureADProvider({
+                clientId: process.env.AD_CLIENT_ID ? process.env.AD_CLIENT_ID : '',
+                clientSecret: process.env.AD_SECRET_VALUE ? process.env.AD_SECRET_VALUE : '',
```
Comment (Xmilofranke) [inline src/authConfig.ts:35]: I think it should also work like this:
```suggestion
                clientSecret: process.env.AD_SECRET_VALUE ?? '',
```

File: src/authConfig.ts [hunk 7]
```diff
@@ -16,37 +19,81 @@ declare global {
...
+                return token;
+            } else if (nowTimeStamp < (token.expires_at as number)) {
+                return token;
+            } else {
+                if(!keycloakEnabled) return token;
+                try {
+                    console.warn('Refreshing access token...');
+                    const newToken = await refreshAccessToken(token);
+                    console.info('Token successfully refreshed.');
+                    return newToken;
```
Comment (Xmilofranke) [inline src/authConfig.ts:62]: ESLint gives a warning here and I think generally we should only log warnings and errors
```suggestion
                    console.warn('Refreshing access token...');
                    return await refreshAccessToken(token);
```

File: src/authConfig.ts [hunk 8]
```diff
@@ -16,37 +19,81 @@ declare global {
...
+                }
+            }
         },
-        cache: {
-            cacheLocation: 'sessionStorage', // This configures where your cache will be stored
-            storeAuthStateInCookie: false, // Set this to "true" if you are having issues on IE11 or Edge
+        async session({ session, token }) {
+            session.accessToken = token.access_token as string;
+            session.idToken = token.id_token as string;
+            return session
```
Comment (Xmilofranke) [inline src/authConfig.ts:72]:
```suggestion
            return session;
```

File: src/authConfig.ts [hunk 9]
```diff
@@ -16,37 +19,81 @@ declare global {
...
-    apiTemplates: {
-        scopes: [`${process.env.NEXT_PUBLIC_APPLICATION_ID_URI}admin.write`], // e.g. api://xxxxxx/access_as_user
-    },
-};
+const refreshAccessToken = async (token: JWT) => {
+    const resp = await fetch(`${process.env.KEYCLOAK_REFRESH_TOKEN_URL}`, {
+        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+        body: new URLSearchParams({
+            client_id: process.env.KEYCLOAK_CLIENT_ID ? process.env.KEYCLOAK_CLIENT_ID : '',
+            client_secret: 'process.env.AZURE_AD_CLIENT_SECRET',
```
Comment (Xmilofranke) [inline src/authConfig.ts:83]: Same question here: Why is this env variable hard coded as a string?

File: src/authConfig.ts (hunk 5 above)
Comment (GailMelanie) [reply src/authConfig.ts:29]: can you comment that in the code? So we don't be suprised about it in a few weeks 😄

File: .env.example [hunk 10]
```diff
@@ -1,3 +1,3 @@
-AD_CLIENT_ID: '<<Azure client ID>>'
-AD_TENANT_ID: '<<Azure tenant ID>>'
-MNESTIX_BACKEND_API_KEY: '<<YOUR_API_KEY>>'
\ No newline at end of file
+AD_SECRET_VALUE: '<<YOUR_SECRET>>'
+MNESTIX_BACKEND_API_KEY: '<<YOUR_API_KEY>>'
+NEXTAUTH_SECRET: '<<YOUR_SECRET>>'
```
Comment (Xmilofranke) [inline .env.example:3]: Is this even used anywhere?

File: docker-compose/compose.only-browser.yml [hunk 11]
```diff
@@ -1,11 +1,11 @@
 services:
   mnestix-browser:
-    image: mnestix/mnestix-browser:latest # Using the pipeline image
+    image: mnestix/mnestix-browser:test-build
```
Comment (Xmilofranke) [inline docker-compose/compose.only-browser.yml:3]: What is this image and why should we use it?

File: src/lib/api/infrastructure.ts [hunk 12]
```diff
@@ -10,46 +10,36 @@ const initializeRequestOptions = async (bearerToken: string, init?: RequestInit)
...
-            );
+            const response = await fetch(url, await initializeRequestOptions(await getBearerToken(), init));
 
             if (response.status !== 401) {
                 return response;
             }
-            // Todo route to login page
-            // await redirectToLoginPage();
-
+            // If not logged in, redirect to sign-in
```
Comment (Xmilofranke) [inline src/lib/api/infrastructure.ts:32]: Is this comment still correct?

File: src/lib/api/infrastructure.ts [hunk 13]
```diff
@@ -10,46 +10,36 @@ const initializeRequestOptions = async (bearerToken: string, init?: RequestInit)
...
             return response;
         },
     };
 };
+
+export const sessionLogOut = async (keycloakEnabled: boolean) => {
+    if(!keycloakEnabled) return;
```
Comment (Xmilofranke) [inline src/lib/api/infrastructure.ts:39]:
```suggestion
    if (!keycloakEnabled) return;
```

File: src/authConfig.ts [hunk 14]
```diff
@@ -16,37 +19,79 @@ declare global {
...
+            } else {
+                if(!keycloakEnabled) return token;
+                try {
+                    console.warn('Refreshing access token...');
+                    return await refreshAccessToken(token);
+                } catch (error) {
+                    console.error('Error refreshing access token', error);
+                    return { ...token, error: 'RefreshAccessTokenError' };
+                }
+            }
```
Comment (Xmilofranke) [inline src/authConfig.ts:65]: You could remove the else block here to reduce nesting
```suggestion
            }
            
            if (!keycloakEnabled) return token;
                            
            try {
                console.warn('Refreshing access token...');
                return await refreshAccessToken(token);
            } catch (error) {
                console.error('Error refreshing access token', error);
                return { ...token, error: 'RefreshAccessTokenError' };
            }
```

File: .env.example (hunk 10 above)
Comment (NilsXitaso) [reply .env.example:3]: This is the template that other users can make their .env file

**Review by GailMelanie — APPROVED** [review]

File: src/authConfig.ts [hunk 15]
```diff
@@ -19,24 +19,43 @@ declare global {
...
 
-const keycloakEnabled = process.env.KEYCLOAK_ENABLED?.toLowerCase() === 'true'.toLowerCase();
+const isEmptyOrWhiteSpace = (input: string | undefined) => {
+    return !input || input.trim() === '';
+}
+
+const keycloakEnabled = (process.env.KEYCLOAK_ENABLED || 'false').toLowerCase() === 'true'.toLowerCase();
+const keycloakLocalUrl = process.env.KEYCLOAK_LOCAL_URL;
+const keycloakIssuer = process.env.KEYCLOAK_ISSUER;
+const serverUrlFromConfig = isEmptyOrWhiteSpace(keycloakLocalUrl) ? keycloakLocalUrl : keycloakIssuer;
```
Comment (Xmilofranke) [inline src/authConfig.ts:29]: Souldn't the logic here be inverted? Right now, the keycloakLocalUrl is used, only when it's empty or a whitespace

File: src/authConfig.ts [hunk 16]
```diff
@@ -19,24 +19,43 @@ declare global {
...
-                clientId: process.env.AD_CLIENT_ID ? process.env.AD_CLIENT_ID : '',
-                clientSecret: process.env.AD_SECRET_VALUE ?? '',
-                tenantId: process.env.AD_TENANT_ID,
-                authorization: { params: { scope: `openid ${process.env.APPLICATION_ID_URI}admin.write` } },
-            })
-        ])
+        ...(keycloakEnabled
+            ? [
+                  KeycloakProvider({
+                      clientId: process.env.KEYCLOAK_CLIENT_ID ? process.env.KEYCLOAK_CLIENT_ID : '',
```
Comment (Xmilofranke) [inline src/authConfig.ts:38]:
```suggestion
                      clientId: process.env.KEYCLOAK_CLIENT_ID ?? '',
```

File: src/authConfig.ts [hunk 17]
```diff
@@ -53,28 +72,27 @@ export const authOptions: AuthOptions = {
...
-                try {
-                    console.warn('Refreshing access token...');
-                    return await refreshAccessToken(token);
-                } catch (error) {
-                    console.error('Error refreshing access token', error);
-                    return { ...token, error: 'RefreshAccessTokenError' };
-                }
+            } 
+            if (!keycloakEnabled) return token;
+            try {
```
Comment (Xmilofranke) [inline src/authConfig.ts:77]:
```suggestion
            } 

            if (!keycloakEnabled) return token;

            try {
```

File: src/lib/api/infrastructure.ts [hunk 18]
```diff
@@ -29,14 +29,13 @@ export const mnestixFetch = (): {
...
-            // If not logged in, redirect to sign-in
             return response;
         },
     };
 };
 
 export const sessionLogOut = async (keycloakEnabled: boolean) => {
-    if(!keycloakEnabled) return;
+    if (!keycloakEnabled) return;
     try {
```
Comment (Xmilofranke) [inline src/lib/api/infrastructure.ts:39]:
```suggestion
    if (!keycloakEnabled) return;

    try {
```

File: package.json [hunk 19]
```diff
@@ -85,7 +86,8 @@
         "docker:test": "docker compose -f compose.yml -f docker-compose/compose.test.yml --profile tests up -d --build && docker compose -f compose.yml -f docker-compose/compose.test.yml attach cypress-test",
         "docker:stop": "docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.test.yml kill",
         "docker:down": "docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.test.yml down",
-        "docker:prune": "docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.test.yml down -v && docker network prune && docker network rm mnestix-network"
+        "docker:prune": "docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.test.yml down -v && docker network prune && docker network rm mnestix-network",
+        "docker:keycloak": "docker compose -f compose.yml -f docker-compose/compose.keycloak.yml up -d"
```
Comment (Xmilofranke) [inline package.json:90]: I think it makes sense to run it in dev mode here, as the keycloak stuff will probably mostly be used in development anyways
```suggestion
        "docker:keycloak": "docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.keycloak.yml up -d"
```

File: README.md [hunk 20]
```diff
@@ -213,20 +238,25 @@ It is possible to change the look and feel by setting a theme color and a person
...
+| `KEYCLOAK_ENABLED`                    | false                       | By default, it is set to false, meaning Keycloak authentication will be disabled, and the default authentication method will be Azure Entra ID. If you set this variable to true, Keycloak authentication will be enabled instead. | optional |
+| `KEYCLOAK_CLIENT_ID`                  | mnestix-browser-client-demo | Configuration variable that specifies the client unique identifier used by your application when connecting to the Keycloak server.                                                                                                | optional |
+| `KEYCLOAK_ISSUER`                     |                             | Configuration variable that specifies the URL of the Keycloak server's issuer endpoint. This endpoint provides the base URL for the Keycloak server that issues tokens and handles authentication requests                         | optional |
```
Comment (Xmilofranke) [inline README.md:257]:
```suggestion
| `KEYCLOAK_ISSUER`                     |                             | Configuration variable that specifies the URL of the Keycloak servers issuer endpoint. This endpoint provides the base URL for the Keycloak server that issues tokens and handles authentication requests                         | optional |
```

File: README.md [hunk 21]
```diff
@@ -521,6 +553,54 @@ The YARP proxy route `/repo/shells` now limits resource and list returns to 104
...
+>
+> For Mnesitx API configuration details, please refer to the API documentation available on [Docker Hub](https://hub.docker.com/r/mnestix/mnestix-api).
+
+
+### Setting Up Keycloak for Docker Development
+
+To start Mnestix along with Keycloak as the authorization server, use one of the following commands:
+
+```sh
+docker compose -f compose.yml -f docker-compose/compose.keycloak.yml up -d
```
Comment (Xmilofranke) [inline README.md:568]:
```suggestion
docker compose -f compose.yml -f docker-compose/compose.dev.yml -f docker-compose/compose.keycloak.yml up -d
```

**Review by Xmilofranke — APPROVED** [review]

**Pull request merged**